pennylane_qiskit/qiskit_device2.py

```diff
@@ -0,0 +1,606 @@
+
+        for kwarg in self._kwargs:
+            program_inputs[kwarg] = self._kwargs.get(kwarg)
+
+        options = {
+            "backend": self.backend.name,
+            "log_level": self.options.environment.log_level,
+            "job_tags": self.options.environment.job_tags,
+            "max_execution_time": self.options.max_execution_time,
+        }
+
+        # Send circuits to the cloud for execution by the circuit-runner program.
+        job = self.service.run(
+            program_id="circuit-runner",
+            options=options,
+            inputs=program_inputs,
+            session_id=session.session_id,
+        )
+        self._current_job = job.result(decoder=RunnerResult)
+
+        results = []
+
+        for index, circuit in enumerate(circuits):
+            self._samples = self.generate_samples(index)
+            res = [
+                mp.process_samples(self._samples, wire_order=self.wires)
+                for mp in circuit.measurements
+            ]
+            single_measurement = len(circuit.measurements) == 1
+            res = res[0] if single_measurement else tuple(res)
+            results.append(res)
+
+        return tuple(results)
+
+    def _execute_sampler(self, circuit, session):
+        """Execution for the Sampler primitive"""
+
+        qcirc = circuit_to_qiskit(circuit, self.num_wires, diagonalize=True, measure=True)
+
+        sampler = Sampler(session=session, options=self.options)
+
+        result = sampler.run(qcirc).result()
+
+        # needs processing function to convert to the correct format for states, and
+        # also handle instances where wires were specified in probs, and for multiple probs measurements
+        # single_measurement = len(circuit.measurements) == 1
+        # res = (res[0], ) if single_measurement else tuple(res)
+
+        return (result.quasi_dists[0],)
+
+    def _execute_estimator(self, circuit, session):
+        # the Estimator primitive takes care of diagonalization and measurements itself,
+        # so diagonalizing gates and measurements are not included in the circuit
+        qcirc = circuit_to_qiskit(circuit, self.num_wires, diagonalize=False, measure=False)
+
+        estimator = Estimator(session=session, options=self.options)
+
+        # split into one call per measurement
+        # could technically be more efficient if there are some observables where we ask
+        # for expectation value and variance on the same observable, but spending time on
+        # that right now feels excessive
+        pauli_observables = [mp_to_pauli(mp, self.num_wires) for mp in circuit.measurements]
+        result = estimator.run([qcirc] * len(pauli_observables), pauli_observables).result()
+        self._current_job = result
+        result = self._process_estimator_job(circuit.measurements, result)
+
+        return result
+
+    @staticmethod
+    def _process_estimator_job(measurements, job_result):
```

Comment:
there can't be other measurements in these tapes because of the preprocessing, right?

Reply:
correct